### PR TITLE
More tolerant NetCDF4 loading - skip un-addable dimensional metadata

### LINKED ISSUE #3644: Netcdf load should workaround unexpected names in 'coordinates'
We encountered a file where loading fails due to a variable referencing as 'coordinates' a dimension that it doesn't actually use.

Simplified example CDL :
```
netcdf minimal_failcase {
dimensions:
                length_scale = 1 ;
                lat = 3 ;
variables:
                float lat(lat) ;
                                lat:standard_name = "latitude" ;
                                lat:units = "degrees_north" ;
                short lst_unc_sys(length_scale) ;
                                lst_unc_sys:long_name = "uncertainty from large-scale systematic errors" ;
                                lst_unc_sys:units = "kelvin" ;
                                lst_unc_sys:coordinates = "lat" ;

data:
  lat = 0, 1, 2;
}
```

**LOADING FAILS** 
with message"ValueError: Missing data dimensions for multi-valued DimCoord 'latitude'"

**Proposal**
Strictly, this is not good CF.
Cf-checker confirms this, saying "INFO: attribute coordinates is being used in a non-standard way".
However, we should probable be capable of gracefully working around this type of problem,
as we do for invalid 'unit' and 'standard_name'.

### PR DESCRIPTION
**Pull request by trexfeathers**
### Description

Closes #3644 (and possibly some others?).

---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

### CODE REVIEW

**Comment by trexfeathers** [comment]
I considered similar catching for the creation of the dimensional metadata instances during NetCDF load, but working out which of the many errors would be sensible to catch would dramatically increase scope. The 16 errors during the addition step were much easier to 'audit'.

**Review by stephenworsley — CHANGES_REQUESTED** [review]
Looks good, that's some gnarly monkeypatch/mock testing but I think it covers things nicely at the unit level.

I think the main thingto add here is an integration test which demonstrates the behaviour of loading a file with these invalid coordinates. This is not entirely obvious from the loading code which is quite complex, though experiments show that these variables tend to be ignored completely rather than become their own cubes. This is in a way preferable to the current behaviour where nothing would be loaded, though it's worth considering future improvements to this behaviour which retain this information in some form. I think such improvements would be quite complicated though, so we can consider this out of scope for the time being (though it would be worth adding in a comment in the integration test stating that the behaviour tested for could be revised in future).

**Review by stephenworsley — CHANGES_REQUESTED** [review]
Just a couple extra details to look at.

File: lib/iris/tests/unit/fileformats/nc_load_rules/helpers/test_build_dimension_coordinate.py [hunk 1]
```diff
@@ -233,6 +241,40 @@ def test_aux_coord_construction(self):
...
+                build_dimension_coordinate(self.engine, self.cf_coord_var)
+            assert self.engine.cube_parts["coordinates"] == []
+
+    def test_auxcoord_not_added(self):
+        # Confirm that a gracefully-created auxiliary coord will also be
+        #  skipped if a CannotAddError is raised when attempting to add.
+        def mock_add_aux_coord(_, __):
+            raise CannotAddError("foo")
+
+        self.monkeypatch.setattr(
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/tests/unit/fileformats/nc_load_rules/helpers/test_build_dimension_coordinate.py]: c1a84f5205667b620afb342d35343f81eb5212ea

File: lib/iris/tests/unit/fileformats/nc_load_rules/helpers/test_build_dimension_coordinate.py [hunk 2]
```diff
@@ -233,6 +241,40 @@ def test_aux_coord_construction(self):
                 warnings.warn.call_args[0][0],
             )
 
+    def test_dimcoord_not_added(self):
+        # Confirm that the coord will be skipped if a CannotAddError is raised
+        #  when attempting to add.
+        def mock_add_dim_coord(_, __):
+            raise CannotAddError("foo")
+
+        self.monkeypatch.setattr(
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/tests/unit/fileformats/nc_load_rules/helpers/test_build_dimension_coordinate.py]: c1a84f5205667b620afb342d35343f81eb5212ea

File: lib/iris/tests/unit/fileformats/nc_load_rules/helpers/test_build_auxiliary_coordinate.py [hunk 3]
```diff
@@ -305,6 +312,20 @@ def test_aux_coord_construction(self):
     def test_aux_coord_construction__climatology(self):
         self.check_case_aux_coord_construction(climatology=True)
 
+    def test_not_added(self):
+        # Confirm that the coord will be skipped if a CannotAddError is raised
+        #  when attempting to add.
+        def mock_add_aux_coord(_, __):
+            raise CannotAddError("foo")
+
+        self.monkeypatch.setattr(
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/tests/unit/fileformats/nc_load_rules/helpers/test_build_auxiliary_coordinate.py]: c1a84f5205667b620afb342d35343f81eb5212ea

File: docs/src/whatsnew/latest.rst [hunk 4]
```diff
@@ -54,6 +54,12 @@ This document explains the changes made to Iris for this release
    both dim and aux coords of the same type e.g. ``longitude`` and ``grid_longitude``.
    (:issue:`3916`, :pull:`5029`).
 
+#. `@trexfeathers`_ made NetCDF loading more tolerant by enabling skipping of
+   :class:`~iris.coords.DimCoord`\s, :class:`~iris.coords.AuxCoord`\s,
+   :class:`~iris.coords.CellMeasure`\s and
+   :class:`~iris.coords.AncillaryVariable`\s if they cannot be added to a
+   :class:`~iris.cube.Cube` (e.g. due to CF non-compliance). (:pull:`5054`)
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply docs/src/whatsnew/latest.rst]: e3a28969c592c83b8b1a92c7e6ec864eef075bd8

Good idea. But I'm pleased to say this change in error class avoids being a breaking change, as `CannotAddError` subclasses the original `ValueError`.

**Review by stephenworsley — APPROVED** [review]
Looks good!

**Pull request merged**